### PR TITLE
LibGfx: Drop tags of unknown type instead of writing invalid icc files

### PR DESCRIPTION
**Pull request by nico**
We could make UnknownTagData hold on to undecoded, raw input data and write that back out when serializing. But for now, we don't.

On the flipside, this _does_ write unknown tags that have known types. We could have a mode where we drop unknown tags with known types. But for now, we don't have that either.

With this, we can for example reencode
/Library/ColorSync/Profiles/WebSafeColors.icc and icc (and other tools) can dump the output icc file. The 'ncpi' tag with type 'ncpi' is dropped while writing it, while the unknown tag 'dscm' with known type 'mluc' is written to the output. (That file is a v2 file, so 'desc' has to have type 'desc' instead of type 'mluc' which it would have in v4 files -- 'dscm' emulates an 'mluc' description in v2 files.)

### CODE REVIEW

File: Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp [hunk 1]
```diff
@@ -569,7 +569,7 @@ static ErrorOr<ByteBuffer> encode_xyz(XYZTagData const& tag_data)
     return bytes;
 }
 
-static ErrorOr<ByteBuffer> encode_tag_data(TagData const& tag_data)
+static Optional<ErrorOr<ByteBuffer>> encode_tag_data(TagData const& tag_data)
```
Comment (linusg) [inline Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp:572]: `ErrorOr<Optional<ByteBuffer>>` please

File: Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp [hunk 2]
```diff
@@ -623,29 +620,37 @@ static ErrorOr<Vector<ByteBuffer>> encode_tag_datas(Profile const& profile, Hash
         if (tag_data_map.contains(tag_data.ptr()))
             return {};
 
-        tag_data_bytes.append(TRY(encode_tag_data(tag_data)));
+        auto encoded_tag_data = encode_tag_data(tag_data);
+        if (!encoded_tag_data.has_value())
+            return {};
+
+        tag_data_bytes.append(TRY(encoded_tag_data.value()));
```
Comment (linusg) [inline Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp:627]: and then this becomes

```cpp
auto encoded_tag_data = TRY(encode_tag_data(tag_data));
if (!encoded_tag_data.has_value())
    return {};
tag_data_bytes.append(encoded_tag_data.release_value());
```

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp (hunk 1 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp:572]: Isn't that something else? An `ErrorOr<Optional<ByteBuffer>>` is an error or maybe a byte buffer, while this maybe make the attempt to create a byte buffer, which can fail.

At least in my head those are two different things!

(Or is this something from the coding style document I had missed?)

File: Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp (hunk 2 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp]: Done, but as said above, it's not clearly better to me, so if you have some words on why it's better it'd be appreciated :)

Comment (nico) [reply Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp]: (The `value()` -> `release_value()` suggestion is a good one, I understand that part.)

File: Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp (hunk 1 above)
Comment (linusg) [reply Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp]: I wouldn't assign different meaning to them, both are "fallible function that may or may not return a byte buffer".
Nested `ErrorOr` is just pretty unusual and it seems nicer to just `TRY` that away upfront, even if it technically doesn't even apply in the `Optional` case.

There's lots of precedent for `ErrorOr<Optional<T>>`: https://github.com/search?q=repo%3ASerenityOS%2Fserenity%20%22ErrorOr%3COptional%22&type=code

But AFAICT none for `Optional<ErrorOr<T>>`: https://github.com/search?q=repo%3ASerenityOS%2Fserenity+%22Optional%3CErrorOr%22&type=code

Comment (linusg) [reply Userland/Libraries/LibGfx/ICC/BinaryWriter.cpp]: I guess the TL;DR is "arbitrary unwritten style preference" 🙃

**Pull request merged**